python/lsst/daf/butler/core/logging.py

```diff
@@ -94,3 +105,270 @@ def MDCRemove(cls, key: str) -> None:
+            if line:  # Filter out blank lines.
+                records.append(ButlerLogRecord.parse_raw(line))
+
+        return cls.from_records(records)
+
+    @property
+    def log_format(self) -> str:
+        if self._log_format is None:
+            return _LONG_LOG_FORMAT
+        return self._log_format
+
+    @log_format.setter
+    def log_format(self, format: str) -> None:
+        self._log_format = format
+
+    def __len__(self) -> int:
+        return len(self.__root__)
+
+    # The signature does not match the one in BaseModel but that is okay
+    # if __root__ is being used.
+    # See https://pydantic-docs.helpmanual.io/usage/models/#custom-root-types
+    def __iter__(self) -> Iterator[ButlerLogRecord]:  # type: ignore
+        return iter(self.__root__)
+
+    def __setitem__(self, index: int, value: Record) -> None:
+        self.__root__[index] = self._validate_record(value)
+
+    def __getitem__(self, index: Union[slice, int]) -> Union[ButlerLogRecords, ButlerLogRecord]:
+        # Handles slices and returns a new collection in that
+        # case.
+        item = self.__root__[index]
+        if isinstance(item, list):
+            return type(self)(__root__=item)
+        else:
+            return item
+
+    def __reversed__(self) -> Iterator[ButlerLogRecord]:
+        return self.__root__.__reversed__()
+
+    def __delitem__(self, index: int) -> None:
+        del self.__root__[index]
+
+    def __str__(self) -> str:
+        # Ensure that every record uses the same format string.
+        return "\n".join(record.format(self.log_format) for record in self.__root__)
+
+    def _validate_record(self, record: Record) -> ButlerLogRecord:
+        if isinstance(record, ButlerLogRecord):
+            pass
+        elif isinstance(record, LogRecord):
+            record = ButlerLogRecord.from_record(record)
+        else:
+            raise ValidationError(f"Can only not append item of type {type(record)}")
+        return record
+
+    def insert(self, index: int, value: Record) -> None:
+        self.__root__.insert(index, self._validate_record(value))
+
+    def append(self, value: Record) -> None:
+        value = self._validate_record(value)
+        self.__root__.append(value)
+
+    def clear(self) -> None:
+        self.__root__.clear()
+
+    def extend(self, records: Iterable[Record]) -> None:
+        self.__root__.extend(self._validate_record(record) for record in records)
+
+    def pop(self, index: int = -1) -> ButlerLogRecord:
+        return self.__root__.pop(index)
+
+    def reverse(self) -> None:
+        self.__root__.reverse()
+
+
+class ButlerLogRecordHandler(StreamHandler):
+    """Python log handler that accumulates records.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.records = ButlerLogRecords(__root__=[])
+
+    def emit(self, record: LogRecord) -> None:
+        self.records.append(record)
+
+
+class JsonFormatter(Formatter):
```

Comment:
There is another class with the same name in this package, maybe call this `JsonLogFormatter`?